core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java

```diff
@@ -0,0 +1,444 @@
+                case EDGES_CONDENSED: {
+                    int min = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    int max = Byte.toUnsignedInt(nodes.get(movingOffset++));
+
+                    if (next < min) {
+                        return takeFirstValueOnLeft(nodes.getLong(movingOffset));
+                    } else if (next > max) {
+                        return takeLastValueOnRight(nodeOffset) + 1;
+                    }
+
+                    int index = next - min;
+                    movingOffset = nodes.getLong(movingOffset + index * Long.BYTES);
+                    break;
+                }
+                case EDGES_NONE:
+                    return maybeValue + 1;
+            }
+        }
+    }
+
+    private int indexOfLessThan(@NotNull final BufferIterator query, final boolean orEquals) {
+        assert keysCount > 0;
+
+        long movingOffset = 0L;
+        long nodeOffset;
+
+        while (true) {
+            final int metadata = Byte.toUnsignedInt(nodes.get(movingOffset));
+            nodeOffset = movingOffset;
+            movingOffset += Byte.BYTES;
+
+            if (hasPrefix(metadata)) { // there is an prefix
+                final int prefixSize = nodes.getInt(movingOffset);
+                movingOffset += Integer.BYTES;
+                final BufferIterator prefix = new BufferIterator(nodes, movingOffset, prefixSize);
+                final int prefixCompare = query.compareToPrefix(prefix);
+                if (prefixCompare > 0) { // query > node
+                    return takeLastValueOnRight(nodeOffset);
+                } else if (prefixCompare < 0) { // prefix > query
+                    return takeFirstValueOnLeft(nodeOffset) - 1;
+                }
+                movingOffset += prefixSize * Byte.BYTES;
+            }
+
+            int maybeValue = NOT_FOUND;
+            if (hasValue(metadata)) { // there is a value
+                maybeValue = nodes.getInt(movingOffset);
+                movingOffset += Integer.BYTES;
+            }
+
+            if (!query.hasNext()) {
+                if (maybeValue != NOT_FOUND) {
+                    return orEquals ? maybeValue : maybeValue - 1;
+                }
+                return takeFirstValueOnLeft(nodeOffset) - 1;
+            }
+
+            final int next = query.next();
+            switch (edgeType(metadata)) {
+                case EDGES_SINGLE:
+                    int key = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    if (next > key) {
+                        return takeLastValueOnRight(nodes.getLong(movingOffset));
+                    } else if (next < key) {
+                        return takeFirstValueOnLeft(nodes.getLong(movingOffset)) - 1;
+                    }
+                    movingOffset = nodes.getLong(movingOffset);
+                    break;
+                case EDGES_BITSET: {
+                    int min = Byte.toUnsignedInt(nodes.get(movingOffset++));
```

Comment:
I am not quite sure, but is it possible to refactor this extremely similar code into separate `private` methods?

Reply:
I think it would be hard without losing performance - that line generates value and also moves an external buffer pointer.

That code is written in while loop with moving pointer to achieve stack safety - that's some tail recursion emulation :)

Reply:
Holy, shit...